autopilot/contractor/contractor.go

```diff
@@ -189,714 +191,271 @@ func (c *Contractor) Close() error {
 	return nil
 }
 
-func canSkipContractMaintenance(ctx context.Context, cfg api.ContractsConfig) (string, bool) {
-	select {
-	case <-ctx.Done():
-		return "", true
-	default:
-	}
-
-	// no maintenance if no hosts are requested
-	//
-	// NOTE: this is an important check because we assume Contracts.Amount is
-	// not zero in several places
-	if cfg.Amount == 0 {
-		return "contracts is set to zero, skipping contract maintenance", true
-	}
-
-	// no maintenance if no allowance was set
-	if cfg.Allowance.IsZero() {
-		return "allowance is set to zero, skipping contract maintenance", true
-	}
-
-	// no maintenance if no period was set
-	if cfg.Period == 0 {
-		return "period is set to zero, skipping contract maintenance", true
+func (c *Contractor) PerformContractMaintenance(ctx context.Context, w Worker, state *MaintenanceState) (bool, error) {
```

Comment:
My recommendation when reviewing this is to just check it out locally and then start here. Everything that changed happens in here.

Reply:
I was going to open a PR for this today, but seeing as you have this huge `contractor` PR already perhaps you can tackle the following: we don't ever call `Close` on the contractor and the implementation of `Close` should probably also use a waitgroup to properly allow maintenance to finish gracefully.